docs/design/resource-states.md

```diff
@@ -0,0 +1,221 @@
+| WaitingForOwner        | Warning                         | The `owner` of this resource cannot be found in Kubernetes. Progress is blocked until the `owner` is created.                                                             |
+| Deleting               | Info                            | The resource is being deleted.                                                                                                                                            |
+| Succeeded              | None (""/empty)                 | The `spec` has successfully been applied. No additional changes are being attempted at this time.                                                                         |
+
+There are no failure conditions (`Severity = Error`) specified here because currently all fatal errors come directly from Azure. When an error response is received from Azure, the `Code` from Azure is set as the `Reason`, and the `Message` from Azure
+is set as the `Message`.
+
+#### `Severity` meaning in the context of the `Ready` condition
+
+* `Info` means that everything is proceeded as expected. This is the "happy path".
+* `Warning` means that something is wrong, but we haven't given up. The resource will continue to be reconciled until we can progress past the cause of the `Warning`. Customers should examine the `Warning` as some may be due to the operator waiting for action from them, as in the case of `Reason=WaitingForOwner, Severity=Warning`. Others may be due to transient unavailability in Azure, as in the case of `Reason=InternalServerError, Severity=Warning`.
+* `Error` means that we were unable to reconcile the resource successfully. A `Ready` condition with `Status=False, Reason=<something>, Severity=Error` is no longer attempting to be reconciled. The user must make an update to the resource to fix one or more configuration problems.
+
+
+## Reference
+1. [Article on conditions and status reporting in Kubernetes](https://dev.to/maelvls/what-the-heck-are-kubernetes-conditions-for-4je7). Note that this article is a bit old as Cluster API uses conditions now, but it gives a great overview of the topic.
+2. [Reccomended shape of Conditions](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions)
+3. [Update condition guidance](https://github.com/kubernetes/community/pull/4521) - and the actual guidance itself is [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)
+4. [More condition reccomendations](https://github.com/kubernetes-sigs/cluster-api/issues/1658#issuecomment-584794287)
+
+## Open questions
+1. There are some small differences between this proposal and what CAPI is doing
+   * CAPI says `Reason` is optional, but we're making it required.
```

Comment:
I believe the _reason_ `Reason` (😄 ) is optional is that it should only be set when the condition is False. There is no Reason on True Conditions

Reply:
I guess the official recommendations do say it should be required though

> reason is required and must not be empty. The actor setting the value should always describe why the condition is the way it is, even if that value is "unknown unknowns". No other actor has the information to make a better choice.

need to go back and check why that's not the case in CAPI

Reply:
Given that there are infinitely many failure cases and really only one success case, it feels "ok" to give a (somewhat useless) `Reason` like `Success` for the one success case, as it then saves everybody from then having think about an empty reason.

With that said the guidance isn't really crystal clear as `Reason` is clearly optional [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties).

Reply:
Got a better answer on the CAPI slack: https://kubernetes.slack.com/archives/C8TSNPY4T/p1625615995097600

In case you can't open the link:


> cecile Yesterday at 4:59 PM
Does anyone know/remember why CAPI conditions have Reason as optional? The KEP from apimachinery says it should be required:
 reason is required and must not be empty. The actor setting the value should always describe why the condition is the way it is, even if that value is “unknown unknowns”. No other actor has the information to make a better choice.

> fabrizio.pandini  6 hours ago
Umhh, this comment does not match what the API types are defining few lines above in the doc
        // The reason for the condition’s last transition in CamelCase.
	// The specific API may choose whether or not this field is considered a guaranteed API.
	// This field may not be empty.

> fabrizio.pandini  6 hours ago
I personally don’t think that always adding a reason to a condition makes sense,
thing e.g,. when a condition signals everything is ok, like e.g. Machine Ready true (adding a reason in this case will be just redundant)
Instead, I think that adding a reason in case the condition are signalling a problem makes much more sense

Reply:
I'm not entirely sure I follow this answer. It seems to say:
1. Actually each specific API can choose if `Reason` is required or not.
2. Fabrizio (and a few others who thumbs up-ed his answer) don't see much reason to include `Reason` when the condition signals OK.

I _agree_ with point 2... but it's a really minor quibble. I can't imagine people getting confused by a reason that was included in the success case, and always including reason has the advantage that you can make a nice simple statement about reason: "It's always set". More importantly than that though, it matches the KEP. The other places where the CAPI diverged from the KEP were explicitly [called out](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200506-conditions.md#kubernetes-conditions).

@porges, @theunrepentantgeek, @babbageclunk - you have thoughts on this topic?

Reply:
A reason like `ReconciliationComplete` would make it clear that the reason (heh) `Ready` is now **true** is that we've done all our stuff. Removes ambiguity about how we got here.

Reply:
I can see that Reason is redundant when Ready is true (it's like that Tolstoy quote about happy families), but I think the simplicity of having it always present is worth that wrinkle.